src/run.py

```diff
@@ -358,18 +359,36 @@ def review_pr(self, pr: GitPullRequest, pr_url: str, run_state: RunState):
             # Votes were converted when the config was loaded.
             assert vote is None or isinstance(vote, int), f"Vote must be an integer. Got: {vote} with type: {type(vote)}"
             # Can't vote on a draft.
-            if not pr.is_draft and is_vote_allowed(reviewer.vote, vote):
-                assert vote is not None
-                reviewer.vote = vote
-                vote_str = map_int_vote(vote)
-                if not is_dry_run:
-                    self.logger.info("SETTING VOTE: '%s'\nTitle: \"%s\"\nBy %s (%s)\n%s", vote_str,
-                                     pr.title, pr_author.display_name, pr_author.unique_name, pr_url)
-                    self.git_client.create_pull_request_reviewer(
-                        reviewer, repository_id, pr.pull_request_id, reviewer_id=user_id, project=project)
-                else:
-                    self.logger.info("Would vote: '%s'\nTitle: \"%s\"\nBy %s (%s)\n%s", vote_str,
-                                     pr.title, pr_author.display_name, pr_author.unique_name, pr_url)
+            if not pr.is_draft:
+                is_allowed = is_vote_allowed(reviewer.vote, vote)
+                # If the new vote is the same as the current vote, don't try to set it again, but do count the rule as
```

Comment:
I think the logic here makes sense, but maybe the setting should change, maybe what we want is something like `allow_less_strict_vote`?

Reply:
Yes, I'll add docs in the readme.  (I didn't do that initially because I didn't see docs for `reset_votes_after_changes` either.)

This feature isn't really designed to be used in conjunction with manual voting, so the UX might not be ideal in that scenario.  I'll clarify in the readme what "if no rule votes" means.

I don't want to reference votes being less strict in the name because that's not really what this feature is about.  That's more of a consequence of how `is_vote_allowed()` is implemented - which could change in the future.  This feature is about resetting votes that were set in earlier runs/iterations of the Code Reviewer under conditions that no longer apply.  As for the name `allow_less_strict_vote` specifically, it sounds like the vote would only be changed if a new vote is cast, so it wouldn't cover my principal use case, which is resetting the vote to `NO_VOTE`.

Good point about cases where a rule isn't applied because no commits were pushed since the Code Reviewer last checked.  That would indeed break my scenario.  I'll add a test case for this and see if I can come up with a solution.

Reply:
Right, I didn't document `reset_votes_after_changes` cause I didn't really want to advertise it because I wasn't sure if it was a good feature to have. I think that one and this new one are good features to have. I think for this setting, it's important to document what was discussed in #52 that it's mainly made for when votes are only done using the script, i.e., for a bot that is only used by this script and not when used to impersonate a person that might also vote manually.
I think the case about already rejecting might need to be checked. Sorry, I didn't go through the tests thoroughly, do we check that?
If the bot is already rejecting, but the PR changed and now the strictest vote is to wait,
`is_allowed = False`
`has_any_rule_voted = False`
So it looks like we would do `reviewer.vote = NO_VOTE` below.

My examples were about manual voting, but I think they still apply with automatic voting.
If a bot is already waiting because another rule decided to wait earlier, and now a rule would vote to wait, I wouldn't really consider that as the rule voting in the way that the script worked before, since it would not vote again. Maybe this doesn't change the logic, but it's more about semantics and naming of variables, so not a huge deal.

Reply:
No, there is no test for the rejecting case.  I'll add one.

I make a distinction between a rule voting and the script voting.  A rule can vote regardless of what the vote is currently set to (unless the current vote is checked in the rule's conditions), but the script can decide not to apply the rule's vote to avoid spurious API calls and notifications.  That's admittedly a somewhat nitpicky distinction.

Maybe once we've worked out how to fix the two cases that probably don't work right now, we'll get a clearer picture of what exactly this feature hinges on and consequently what it should be named.

Reply:
I added tests for the two cases. As predicted, they fail at the moment.

I think the one about changing the vote from "reject" to "wait" could potentially be fixed by changing the feature: instead of resetting the vote if no rule votes, just reset the vote before all rules run.
I know we had considered something similar before with rules at the top of the list that would reset the vote under certain conditions. We had concluded that that wouldn't work because it wouldn't override a "wait" or "reject" vote. But that wouldn't be a problem with this new feature because it wouldn't be subject to the `is_vote_allowed()` function.
To avoid spurious API calls and notifications, the reset would only happen in memory first and the API would be called after all rules have finished running (if no rule ended up calling the API). If we're worried about rules throwing exceptions before the API can be called to reset the vote, we could put the API call in a try-finally block, but I'll leave that up to you if you think it's worth it. I think if the script is throwing an exception, it's not that big a deal if it leaves things in a weird inconsistent state.

Re. the issue with caching the latest commit seen, I think we could also cache which rules (or maybe which diff patterns) matched that commit. Then we can basically re-run those rules without having to download all the file diffs again.
Or maybe we could just generally cache the output of those file diff API calls, so we can separate the business logic from the caching logic.

Let me know what you think!